lab-bundle-spec.md

```diff
@@ -158,41 +163,63 @@ environment_resources:
+startup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+startup_script.custom_properties |          | array   | Array of pairs. See below for details.
+cleanup_script.type              |          | string  | The type of cleanup script. Only `deployment_manager` is supported.
+cleanup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+cleanup_script.custom_properties |          | array   | Array of pairs. See below for details.
+ssh_key_user                     |          | string  | If this project should use a user's SSH key, the id of that user.
 
 ```yml
 - type: gcp_project
   id: secondary_project
-  fleet: gcpfree
-  dm_template:
-    script: deployment_manager
+  variant: gcpfree
+  startup_script:
+    type: deployment_manager
+    path: dm_startup.zip
     custom_properties:
       - key: userNameWindows
         value: student
+  cleanup_script:
+    type: deployment_manager
+    path: dm_cleanup.zip
+    custom_properties:
+      - key: primary_project_zone
+        value: my_primary_project.zone
 ```
 
-> **NOTE:** Not all GCP fleet names are supported.
->
-> The existing concept of Qwiklabs' Fleets does not have a single analog in
-> content bundles. Notice that some fleet types map to resource types (e.g.
-> `gsuite_multi_tenant` fleet is now the `gsuite-domain` resource type).
-> However other fleets are "variants" of the same resource type (e.g.
-> "free", "ASL", and "standard" GCP projects). Therefore, not all fleet names that
-> an experienced Qwiklabs author may be familiar with, are allowed in the `fleet`
-> field.
+> **NOTE:** The existing concept of Qwiklabs' Fleets does not have a single
+> analog in content bundles.
 >
-> Presently, authors are allowed to specify one of the following fleet types for
-> a gcp_project:
-> - gcpd [default]
-> - gcpfree
-> - gcpasl
->
-> Future versions of the Content Bundle spec may use different terminology for
-> resource_type variations to avoid conflation.
+> Some fleet types map to resource types (e.g. `gsuite_multi_tenant` fleet is
+> now the `gsuite-domain` resource type), while other fleets are allowed as
+> "variants" of `gcp_project` (see below).
+
+###### Variants
+
+The allowed variants are:
+- gcpd [default]
+- gcpfree
+- gcpasl
+
+
+###### Custom properties
+
+attribute | required | type   | notes
+----------| -------- | ------ | --------------------------------------
+key       | ✓        | string | How the property will be referenced within the script.
+value     |          | string | A value to be passed into the script.
+reference |          | string | A reference to a value, in the form `[RESOURCE].[PROPERTY]`, which will be obtained from the running resource and passed into the script.
+
+The valid `reference`s are:
+- [USER_ID].username
+- [USER_ID].password
+- [PROJECT_ID].project_id
+- [PROJECT_ID].zone
```

Comment:
Nit: how about `default_zone` to draw more attention to the fact that this is the default place a student will create resources but is by no means the only place they can do so?

Reply:
Done

Reply:
I'm _very slightly_ concerned about the meaning of "default" here.

There are two "defaults" in operation here:

1. The default zone of the `Lab` as specified by the lab author.
2. The default zone of the `LabInstance` as determined by the context. For instance, a classroom in Japan can override the default zone of the lab to use a closer region/zone.

.... but, unlike the Lab model in the website code base, ContentBundle does not support lab level `default_zone`. So either the point is moot or we need to re-add it.

Reply:
Relatedly, I'm concerned that people will try taking a lab in a zone that does not support all of the features they need. Practically, 90% of labs happen in us-central1 which is one of the first places to get new features so this hasn't come up much in the past.

See https://cloud.google.com/about/locations/?tab=americas for a breakdown of features by region. But, long story short, it is not always safe to attempt a lab in a region that is geographically close to the lab taker.

Instead, what if labs specified a list of `valid_regions` and Core Platform handled picking a region that was close to the lab taker? Not specifying a list would imply the lab is available in all regions.